torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -435,11 +439,61 @@ std::vector<at::Tensor> ProcessGroupGloo::AsyncWork::result() {
       "Work needs to be completed before calling result(). "
       "Should call wait() before result().");
   TORCH_CHECK(
-      outputTensors_.size() <= 1, "work result does not support list of lists.");
+      outputTensors_.size() <= 1,
+      "work result does not support list of lists, use .getFuture() and value()");
   return outputTensors_.size() == 0 ? std::vector<at::Tensor>()
                                     : outputTensors_.at(0);
 }
 
+c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupGloo::AsyncWork::
+    getFuture() {
+  return future_;
+}
+
+namespace {
+c10::intrusive_ptr<c10::ivalue::Future> CreateFutureForOutput(
+    const std::vector<std::vector<at::Tensor>>& outputTensors) {
+  if (outputTensors.size() > 1) {
+    return c10::make_intrusive<c10::ivalue::Future>(
+        c10::ListType::create(c10::ListType::create(c10::TensorType::get())));
+  }
+  return c10::make_intrusive<c10::ivalue::Future>(
+      c10::ListType::create(c10::TensorType::get()));
+}
+
+void ReturnFutureWithOutput(
+    c10::intrusive_ptr<c10::ivalue::Future> future,
+    const std::vector<std::vector<at::Tensor>>& outputTensors) {
+  if (outputTensors.size() == 0) {
+    future->markCompleted(c10::IValue(std::vector<at::Tensor>()));
+    return;
+  }
+  if (outputTensors.size() > 1) {
+    future->markCompleted(c10::IValue(outputTensors));
+    return;
+  }
+  future->markCompleted(c10::IValue(outputTensors[0]));
+}
+} // namespace
+
+ProcessGroupGloo::AsyncWork::AsyncWork(
+    std::vector<std::vector<at::Tensor>> outputTensors,
+    const char* profilingTitle,
+    const c10::optional<std::vector<at::Tensor>>& inputTensors)
+    : ProcessGroup::Work(-1, OpType::UNKNOWN, profilingTitle, inputTensors),
```

Comment:
(This is prior to this PR): Curious, any reason for using UNKNOWN OpType here? I saw UNKNOWN is also used for recv, but I don't recall the reason for that.

Reply:
I've looked around and seems like optype is only used in NCCL process group for p2p comm at #45873. Maybe there are other uses which I haven't found?

https://github.com/pytorch/pytorch/search?q=retrieveOpType

https://github.com/pytorch/pytorch/search?q=opType_